File: solutions/ocaml/083-reorder-list.ml
(*
  Problem 83: Reorder List (LeetCode 143)
  Difficulty: Med
  Language: OCaml
*)
let reorder_list lst =
  let n = List.length lst in
  let mid = n / 2 in
  let first = List.filteri (fun i _ -> i < mid) lst in
  let second = List.rev (List.filteri (fun i _ -> i >= mid) lst) in
  let rec merge a b = match a, b with
    | [], bs -> bs | as_, [] -> as_
    | ah :: at, bh :: bt -> ah :: bh :: merge at bt
  in merge first second

File: solutions/scala/083-reorder-list.scala
/**
 * Problem 83: Reorder List (LeetCode 143)
 * Difficulty: Med
 * Language: Scala
 */
def reorderList(list: List[Int]): List[Int] = {
  val mid = list.length / 2
  val (first, secondHalf) = list.splitAt(mid)
  val second = secondHalf.reverse
  def merge(a: List[Int], b: List[Int]): List[Int] = (a, b) match {
    case (Nil, bs) => bs
    case (as_, Nil) => as_
    case (ah :: at, bh :: bt) => ah :: bh :: merge(at, bt)
  }
  merge(first, second)
}
